src/scheduler/controller.rs

```diff
@@ -79,76 +72,49 @@ impl<VM: VMBinding> GCController<VM> {
             return true;
         }
 
-        // If all fo the above failed, it means GC has finished.
+        // If all of the above failed, it means GC has finished.
         false
     }
 
-    /// Reset the "all workers parked" state and resume workers.
-    fn reset_and_resume_workers(&mut self) {
-        self.receiver.reset_all_workers_parked();
-        self.scheduler.worker_monitor.notify_work_available(true);
-        debug!("Workers resumed");
-    }
-
     /// Handle the "all workers have parked" event.  Return true if GC is finished.
     fn on_all_workers_parked(&mut self) -> bool {
-        assert!(self.scheduler.all_activated_buckets_are_empty());
+        self.scheduler.assert_all_activated_buckets_are_empty();
 
         let new_work_available = self.find_more_work_for_workers();
 
-        if new_work_available {
-            self.reset_and_resume_workers();
-            // If there is more work to do, GC has not finished.
-            return false;
-        }
-
-        assert!(self.scheduler.all_buckets_empty());
-
-        true
-    }
-
-    /// Process an event. Return true if the GC is finished.
-    fn process_event(&mut self, message: Event<VM>) -> bool {
-        match message {
-            Event::Work(mut work) => {
-                self.execute_coordinator_work(work.as_mut(), true);
-                false
-            }
-            Event::AllParked => self.on_all_workers_parked(),
-        }
+        // GC finishes if there is no new work to do.
+        !new_work_available
     }
 
     /// Coordinate workers to perform GC in response to a GC request.
     pub fn do_gc_until_completion(&mut self) {
         let gc_start = std::time::Instant::now();
-        // Schedule collection.
-        self.execute_coordinator_work(&mut ScheduleCollection, true);
-
-        // Tell GC trigger that GC started - this happens after ScheduleCollection so we
-        // will know what kind of GC this is (e.g. nursery vs mature in gen copy, defrag vs fast in Immix)
-        self.mmtk
-            .plan
-            .base()
-            .gc_trigger
-            .policy
-            .on_gc_start(self.mmtk);
-
-        // React to worker-generated events until finished.
+
+        debug_assert!(
+            self.scheduler.worker_monitor.debug_is_sleeping(),
+            "Workers are still doing work when GC started."
+        );
+
+        // Add a ScheduleCollection work packet.  It is the seed of other work packets.
+        self.scheduler.work_buckets[WorkBucketStage::Unconstrained].add(ScheduleCollection);
```

Comment:
The ultimate goal is removing the coordinator thread, too, so I am trying to as few things in the coordinator thread as possible.

With the `worker_group_state` I introduced, adding a work packet to a bucket will no longer wake up workers if they are in the `Sleeping` state. I think I need to change the `WorkBucket::add` and the `notify_work_available` method so that they do not notify workers if the state is still `Sleeping`.  Otherwise, when they wake up, they will find it is still `Sleeping` state and `wait()` again.

But you are right.  At this time, only one work packet `ScheduleCollection` exists.  It's appropriate to make a special case for `resume_and_wait` so that it wakes up only one worker.